curriculum_algorithm/models.py

```diff
@@ -0,0 +1,122 @@
+    season = models.CharField(max_length=1, choices=POSITION_CHOICES)
+    credit_hours = models.IntegerField(default=0)
+    laboratory = models.ForeignKey('Course', null=True, on_delete=models.SET_NULL, related_name='main_course')
+
+    def __str__(self):
+        return self.course_number
+
+
+class Curriculum(models.Model):
+    name = models.CharField(max_length=50, unique=True)
+    department = models.CharField(max_length=50)
+    courses = models.ManyToManyField(Course, through='CurriculumCourse')
+
+    def __str__(self):
+        return self.name
+
+
+class CurriculumCourse(models.Model):
+    course = models.ForeignKey(Course, on_delete=models.CASCADE)
+    curriculum = models.ForeignKey(Curriculum, on_delete=models.CASCADE)
+    pre_requisites = models.ManyToManyField('self', related_name='bops', symmetrical=False)
+    co_requisites = models.ManyToManyField("self", related_name='enmendated_bops', symmetrical=False)
+    level = models.IntegerField(default=0)
+
+    def add_pre_requisite(self, curriculum_course):
+        self.pre_requisites.add(curriculum_course)
+        self.level = max(self.level, curriculum_course.level) + 1
+        self.save()
+
+    def get_credit_hours(self):
+        return self.course.credit_hours + self.course.laboratory.credit_hours \
+            if self.course.laboratory else self.course.credit_hours
+
+    def __str__(self):
+        return f'{self.curriculum.name}: {self.course.course_number}'
+
+
+class Semester(models.Model):
+    student_plan = models.ForeignKey('StudentPlan', on_delete=models.CASCADE)
+    curriculum_courses = models.ManyToManyField(CurriculumCourse)
+    is_full = models.BooleanField(default=False)
+    past = models.BooleanField(default=False)
+    max_credits = models.IntegerField()
+    position = models.IntegerField(default=0)
+    credit_hours = models.IntegerField(default=0)
+
+    # def __str__(self):
+    #     courses = [course.__str__() + '\n' for course in self.curriculum_courses.all()]
+    #     return ''.join(courses)
+
+    def add_curriculum_course(self, curriculum_course):
+        self.curriculum_courses.add(curriculum_course)
+        if curriculum_course.course.laboratory:
+            self.curriculum_courses.add(curriculum_course)
+        self.credit_hours += curriculum_course.get_credit_hours()
+        self.is_full = self.credit_hours >= self.max_credits
+        self.save()
+
+    def course_valid(self, curriculum_course):
+        if not self.is_full and not self.past \
+                and self.credit_hours + curriculum_course.get_credit_hours() <= self.max_credits:
+            if curriculum_course.course.season == 2:
+                return self.position % 2 == 0
+            elif curriculum_course.course.season == 1:
+                return self.position % 2 == 1
+            else:
+                return True
+
+
+class StudentPlan(models.Model):
+    max_credits = models.IntegerField(default=0)
```

Comment:
Consider PositiveIntegerField or another way to avoid negative numbers in max_credits per semester